### PR TITLE
[SPARK-34119][SQL] Keep necessary stats after partition pruning

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr keep necessary stats after partition pruning.

### Why are the changes needed?

Improve query performance. It will push down aggregate since SPARK-34081 because it can be planed as BroadcastHashJoin. But it lacks column statistics after [`PruneFileSourcePartitions`](https://github.com/apache/spark/blob/d0c83f372b1aa87327eacc16b131c57fccf6c205/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala#L102-L103). Therefore, it will eventually be planned as SortMergeJoin.

Please see the log:
```
join.right.stats: org.apache.spark.sql.catalyst.optimizer.PushDownPredicates: Statistics(sizeInBytes=348.8 KiB, rowCount=1.79E+4)
join.right.stats: org.apache.spark.sql.execution.datasources.PruneFileSourcePartitions: Statistics(sizeInBytes=1414.2 EiB)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and benchmark test

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q14a | 594 | 384
q14b | 600 | 402

This change will not affect the results of `PlanStabilitySuite`, because it does not have partition column.

### CODE REVIEW

**Comment by peter-toth** [comment]
This looks much better than my PR: https://github.com/apache/spark/pull/29132
@wangyum I wonder if you checked other than q14 query improvements?

**Comment by cloud-fan** [comment]
This patch alone doesn't change the query plan, so I think it's fine to skip benchmark as it shouldn't change anything.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
shall we do the same in PruneHiveTablePartitions?

**Comment by wangyum** [comment]
> shall we do the same in PruneHiveTablePartitions?

Yes, PruneHiveTablePartitions also has this issue.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PrunePartitionSuiteBase.scala [hunk 1]
```diff
@@ -17,14 +17,13 @@
...
-import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.StatisticsCollectionTestBase
 import org.apache.spark.sql.catalyst.expressions.{AttributeReference, BinaryOperator, Expression, IsNotNull, Literal}
 import org.apache.spark.sql.execution.{FileSourceScanExec, SparkPlan}
 import org.apache.spark.sql.hive.test.TestHiveSingleton
 import org.apache.spark.sql.internal.SQLConf.ADAPTIVE_EXECUTION_ENABLED
-import org.apache.spark.sql.test.SQLTestUtils
 
-abstract class PrunePartitionSuiteBase extends QueryTest with SQLTestUtils with TestHiveSingleton {
+abstract class PrunePartitionSuiteBase extends StatisticsCollectionTestBase with TestHiveSingleton {
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PrunePartitionSuiteBase.scala:26]: this change to use `checkOptimizedPlanStats` from `StatisticsCollectionTestBase `.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**